### PR TITLE
Provide a fix-it that inserts omitted generic parameters when they can't be inferred

### PR DESCRIPTION
**Pull request by jrose-apple**
This will help when migrating to Swift 3 and Objective-C classes with newly-added generic parameters.

Still to do:
- More tests
- See if there's a way to handle conflicting context

rdar://problem/27087345

### CODE REVIEW

**Comment by jrose-apple** [comment]
@nkcsgexi, @akyrtzi, what do you think about what's there so far? In particular, what do you think about the defaults?

- No constraints, `class X<T>`: `X<Any>()`. Should this be `X<<#T: Any##Any#>>` instead?
- Superclass constraint only, `class X<T: UIView>`: `X<UIView>()`.
- Class-bound only, `class X<T: class>`: `X<AnyObject>()`. Should this be `X<<#T: class##AnyObject#>>` instead?
- Single ObjC protocol, `class X<T: NSCoding>`: `X<NSCoding>()`.
- Multiple protocols, or class+protocols: `class X<T: UIViewController, UITableViewDataSource>`: `X<<#T: UIViewController & UITableViewDataSource#>>`

Stepping back, is it okay to use placeholders at all here? (@DougGregor?) I see we have one other diagnostic with placeholders. Maybe we can change the text output format to special-case them (later).

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

File: lib/Sema/CSDiag.cpp [hunk 1]
```diff
@@ -6411,9 +6419,69 @@ static void noteArchetypeSource(const TypeLoc &loc, ArchetypeType *archetype,
...
   
-  if (FoundDecl)
+  if (FoundDecl) {
     tc.diagnose(FoundDecl, diag::archetype_declared_in_type, archetype,
                 FoundDecl->getDeclaredType());
+  }
+  if (FoundGenericTypeBase && !isa<GenericIdentTypeRepr>(FoundGenericTypeBase)){
+    assert(FoundDecl);
+
+    SmallString<64> genericParamBuf;
```
Comment (slavapestov) [inline lib/Sema/CSDiag.cpp]: Is this in any way similar to diagnoseUnboundGenericType() in TypeCheckType.cpp?

Comment (jrose-apple) [inline lib/Sema/CSDiag.cpp]: Oh, good catch. I'll see if I can factor out common logic here. (I like my implementation better.)

**Comment by nkcsgexi** [comment]
I like the idea of using placeholders in fixits. We have a precedent of that in the fixits for auto-filling protocol conformances skeletons.

**Comment by DougGregor** [comment]
I like the idea of using placeholders. It works very well in Xcode, and is fairly easy to handle in other tools that might want to present Fix-Its to the user.

**Comment by jrose-apple** [comment]
Implementation mostly done, but I still need to write tests.

@swift-ci Please test

**Comment by jrose-apple** [comment]
[Apple](https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx/), [Linux](https://ci.swift.org/view/Pull%20Request/job/swift-PR-Linux/3756/)

**Comment by jrose-apple** [comment]
@nkcsgexi, would you mind reviewing this formally for the Swift 3.0 branch? And would you like me to make sure these fix-its are all enabled in migration mode?

@swift-ci Please test

**Comment by jrose-apple** [comment]
Whoops, a test change snuck in there. Gotta fix that.

**Comment by jrose-apple** [comment]
And it looks like master's messed up at the moment. I'll wait.

**Comment by nkcsgexi** [comment]
Will review. This is a fixit associated with error, right? Isn't that automatically picked up by migrator?

**Comment by jrose-apple** [comment]
I guess also:
- @slavapestov, is this an okay use of GenericTypeParamDecl, or should I be using the generic signature and environment?
- @DougGregor or @rudkx, is this a reasonable way to fish for resolved type parameters?

**Comment by jrose-apple** [comment]
@nkcsgexi The new fix-it is on a note because usually you want a more specific type than the default…but in migration, you might not have a type because in Swift 2.3 it wasn't generic.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test Linux platform

**Comment by jrose-apple** [comment]
macOS full tests passed earlier.

@swift-ci Please smoke test macOS platform

**Comment by jrose-apple** [comment]
^ @aciidb0mb3r, have you seen this SwiftPM failure before? (unrelated to this patch)

**Comment by aciidb0mb3r** [comment]
@jrose-apple unfortunately yes, I disabled a similar test failure here 
https://github.com/apple/swift-package-manager/commit/57c5be1db1c1e12e089dff02241ffbce5722fb0e

will disable all related tests.

**Comment by aciidb0mb3r** [comment]
Disabled here: https://github.com/apple/swift-package-manager/commit/76714d283ea63f4fbb8e7668bcd53d96c54ee102

**Comment by jrose-apple** [comment]
Thanks, Ankit.

@swift-ci Please test Linux platform

File: lib/Sema/CSDiag.cpp [hunk 2]
```diff
@@ -6491,39 +6491,105 @@ void ConstraintSystem::diagnoseFailureForExpr(Expr *expr) {
...
+        [&](const GenericTypeParamDecl *genericParam) -> Type {
+      if (genericArgs.empty())
+        return Type();
+
+      Type preferred = genericArgs[genericParam->getIndex()];
+      if (!preferred || preferred->is<ErrorType>())
+        return Type();
+
+      TypeVariableType *unused;
+      Type maybeFixedType = cs.getFixedTypeRecursive(preferred, unused,
```
Comment (nkcsgexi) [inline lib/Sema/CSDiag.cpp]: I'm a little confused here. If we can get generic argument type, why do we invoke CS to fix it? to remove unsubstituted type variables?

Comment (jrose-apple) [inline lib/Sema/CSDiag.cpp]: This is for cases where we have some but not all of the generic arguments. Suggesting `<Int, Any>` is usually better than `<Any, Any>`.

Comment (nkcsgexi) [inline lib/Sema/CSDiag.cpp]: oh, I see. Mind adding some inline comments?

File: lib/Sema/MiscDiagnostics.cpp [hunk 3]
```diff
@@ -932,6 +932,69 @@ static void diagnoseImplicitSelfUseInClosure(TypeChecker &TC, const Expr *E,
...
+          continue;
+        genericParamText << " & " << proto->getName();
+      }
+      genericParamText << "#>";
+      return;
+    }
+
+    if (protocols.empty()) {
+      genericParamText << "Any";
+      return;
```
Comment (nkcsgexi) [inline lib/Sema/MiscDiagnostics.cpp]: For future proof, can we retrieve "Any" from ASTContext.getAnyDecl()->getName().

Comment (nkcsgexi) [inline lib/Sema/MiscDiagnostics.cpp]: Otherwise, LGTM!

Comment (jrose-apple) [inline lib/Sema/MiscDiagnostics.cpp]: There's no "Any" decl anymore, but I'll switch it to `Context.ID_Any` so it's at least more searchable.

**Comment by jrose-apple** [comment]
Updated for Xi's review comments.

@swift-ci Please smoke test

**Comment by nkcsgexi** [comment]
Thank you!

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
Connectivity issue.

@swift-ci Please smoke test OS X platform

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test OS X platform

**Comment by shahmishal** [comment]
@swift-ci Please smoke test OS X platform

**Comment by jrose-apple** [comment]
Still waiting for feedback from @DougGregor or @rudkx on accessing the constraint system, and @slavapestov on the use of GenericTypeParamDecl.

**Comment by rudkx** [comment]
@jrose-apple I suspect this is safe, but I could not guarantee it.

**Pull request merged**